### PR TITLE
Enable multiple LoRa adapters

### PR DESCRIPTION
**Pull request by drbh**
This PR is a work in progress to add support for mutliple loras to be loaded at startup and then use 0 or 1 adapters in a request by specifying the adapter id. 

## Example usage

download adapter without auto merging
```bash
text-generation-server download-weights predibase/dbpedia
text-generation-server download-weights predibase/customer_support
```

start server with multiple LoRa adapters

```bash
LORA_ADAPTERS=predibase/customer_support,predibase/dbpedia \
text-generation-launcher --model-id mistralai/Mistral-7B-v0.1
```

sending request without adapter id

```bash
curl 127.0.0.1:3000/generate \
    -X POST \
    -H 'Content-Type: application/json' \
    -d '{
    "inputs": "What are 3 unique words that describe you?",
    "parameters": {
    "max_new_tokens": 40
  }
}'
```

```json
{
  "generated_text": "\n\nI’m a very passionate person. I’m very driven. I’m very determined.\n\nWhat is your favorite thing about being a teacher?\n\nI love the fact"
}
```

with first LoRa adapter specified

```bash
curl 127.0.0.1:3000/generate \
    -X POST \
    -H 'Content-Type: application/json' \
    -d '{
    "inputs": "What are 3 unique words that describe you?",
    "parameters": {
    "max_new_tokens": 40,
    "adapter_id": "predibase/customer_support"
  }
}'
```

```json
{
  "generated_text": "\n\nI’m not sure if I can come up with 3 unique words that describe me, but I’ll try.\n\n1. Creative\n2. Funny\n3."
}
```

with second LoRa adapter specified

```bash
curl 127.0.0.1:3000/generate \
    -X POST \
    -H 'Content-Type: application/json' \
    -d '{
    "inputs": "You are given the title and the body of an article below. Please determine the type of the article.### Title: Great White Whale\n\n### Body: Great White Whale is the debut album by the Canadian rock band Secret and Whisper. The album was in the works for about a year and was released on February 12 2008.",                                                                                             
    "parameters": {
    "max_new_tokens": 40,
    "adapter_id": "predibase/dbpedia"
  }
}'
```

```json
{
  "generated_text": "8"
}
```

### CODE REVIEW

**Comment by flozi00** [comment]
Hey guys, as I am switched to lorax and started contributing there a lot after the first license change I am happy to see the PR got opened

I would be happy if you are open for some questions and discussion about this.
What do you think about following the lorax style for the api so it could be a drop in replacement, especially for the openai endpoint?
Furthermore are you open for some kernel optimizations like punica, used in lorax and vllm, to minimize the overhead and enable efficient batching of different adapters ?

I'd would be happy to contribute here too.

**Comment by drbh** [comment]
hi @flozi00 thanks for the feedback! can you share more about the lorax style api? I see that in lorax you can specify the adapter via the `model` field in the chat endpoint, is that the feature you're referring to? regarding kernel optimizations, yes we are very interested and plan on using optimized kernels/punica (planning on diving into this specific thing tomorrow/this week). Any suggestions/contributions/patches related to this are always appreciated!

**Comment by drbh** [comment]
update: 

This PR's implementation has been updated to align with the great work done by the [lorax](https://github.com/predibase/lorax) team. This implementation tries to use the same layers when possible and only diverges to work with TGI's recent updates/improvements and limits lora to loading at startup. Current changes allow weights to be loaded similar to Lorax, however there are still issues with generation to be resolved, and other refactors

**Comment by flozi00** [comment]
Looks like you are successfully adopting the lorax code
Please tell me if you need any help with this feature

**Comment by drbh** [comment]
@flozi00 generation with loras is stable now. Just focusing on the rebase then refactors now. And thank you 🙂 a review once the PR is ready would be super helpful!

**Comment by tgaddair** [comment]
Thanks for the shoutout in the docs! It's quite interesting to see things come full circle, maybe we should chat about merging our projects.

**Comment by drbh** [comment]
of course @tgaddair thank you for the awesome work! thats an interesting idea and we are always aiming to improve TGI. We appreciate any contributions/discussions about features that may be helpful to our users

**Comment by flozi00** [comment]
> Thanks for the shoutout in the docs! It's quite interesting to see things come full circle, maybe we should chat about merging our projects.

I'd love to migrate to tgi again 👍  And of course trying to contribute here too

**Comment by xiadingZ** [comment]
hello, how to load lora adapter from local path

**Comment by drbh** [comment]
hi @xiadingZ in this PR lora adapters are loaded from the `HUGGINGFACE_HUB_CACHE` directory. Once you've downloaded the lora locally, you can specify the id like `LORA_ADAPTERS=predibase/customer_support` which will use the local lora model.

once this initial lora work is merged we'll follow up with other improvement such as easier ways to specify lora path, and etc

**Comment by xiadingZ** [comment]
> hi @xiadingZ in this PR lora adapters are loaded from the `HUGGINGFACE_HUB_CACHE` directory. Once you've downloaded the lora locally, you can specify the id like `LORA_ADAPTERS=predibase/customer_support` which will use the local lora model.
> 
> once this initial lora work is merged we'll follow up with other improvement such as easier ways to specify lora path, and etc

Hi, @drbh I can try your methods with downloaded lora.  But I have a lora adapter trained locally. It doesn't have a directory  structure such as `blobs,  refs,  snapshots` from huggingface.  how can I place it in HUGGINGFACE_HUB_CACHE and load it?

I set HUGGINGFACE_HUB_CACHE as `/root/lora_adapters` and my directory structure is:

**Comment by xiadingZ** [comment]
> hi @xiadingZ in this PR lora adapters are loaded from the `HUGGINGFACE_HUB_CACHE` directory. Once you've downloaded the lora locally, you can specify the id like `LORA_ADAPTERS=predibase/customer_support` which will use the local lora model.
> 
> once this initial lora work is merged we'll follow up with other improvement such as easier ways to specify lora path, and etc

Hi, @drbh I can try your methods with downloaded lora.  But I have a lora adapter trained locally. It doesn't have a directory  structure such as `blobs,  refs,  snapshots` from huggingface.  how can I place it in HUGGINGFACE_HUB_CACHE and load it?

I set HUGGINGFACE_HUB_CACHE as `/root/lora_adapters` and my directory structure is:

File: server/text_generation_server/utils/adapter.py
Comment (danieldk) [inline server/text_generation_server/utils/adapter.py:23]: Could be a dataclass. Also adds typing.

File: server/text_generation_server/utils/adapter.py
Comment (danieldk) [inline server/text_generation_server/utils/adapter.py:34]: Could be a dataclass.

File: server/text_generation_server/utils/adapter.py
Comment (danieldk) [inline server/text_generation_server/utils/adapter.py:68]: Could be a dataclass.

File: server/text_generation_server/utils/adapter.py
Comment (danieldk) [inline server/text_generation_server/utils/adapter.py:126]:
```suggestion
            # Avoid execution latency caused by the network connection retrying for AutoConfig.from_pretrained(None)
```

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (danieldk) [inline server/text_generation_server/models/custom_modeling/flash_llama_modeling.py:261]: Are the LoRa parameters also separate in the Phi3 case, since Phi3 uses a single gate+up projection?

File: server/Makefile-lorax-punica
Comment (drbh) [reply server/Makefile-lorax-punica:1]: that would be preferred however I'm not sure lorax are built and available, their CI process does not seem to compile a prebuilt container/kernels. If there is a way to avoid this build step I'd love to 🙂

File: server/text_generation_server/adapters/__init__.py
Comment (drbh) [reply server/text_generation_server/adapters/__init__.py]: removed since unused

File: server/text_generation_server/adapters/__init__.py
Comment (drbh) [reply server/text_generation_server/adapters/__init__.py]: removed

File: server/text_generation_server/adapters/config.py
Comment (drbh) [reply server/text_generation_server/adapters/config.py]: agreed, upgraded to a dataclass thanks

File: server/text_generation_server/adapters/config.py
Comment (drbh) [reply server/text_generation_server/adapters/config.py]: updated to `Dict[int, AdapterWeights]`

File: server/text_generation_server/adapters/config.py
Comment (drbh) [reply server/text_generation_server/adapters/config.py]: updated

File: server/text_generation_server/adapters/lora.py
Comment (drbh) [reply server/text_generation_server/adapters/lora.py]: updated

File: server/text_generation_server/utils/adapter.py
Comment (drbh) [reply server/text_generation_server/utils/adapter.py:28]: updated

File: server/text_generation_server/utils/adapter.py
Comment (drbh) [reply server/text_generation_server/utils/adapter.py:37]: updated

File: server/text_generation_server/utils/adapter.py
Comment (drbh) [reply server/text_generation_server/utils/adapter.py:71]: updated

File: server/text_generation_server/utils/adapter.py
Comment (drbh) [reply server/text_generation_server/utils/adapter.py]: updated

File: server/text_generation_server/utils/peft.py
Comment (drbh) [reply server/text_generation_server/utils/peft.py]: updated

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py]: great catch, since the `adapter_target_to_layer` is not implemented for flash_phi I believe this could cause and issue. 

I've updated the PR to avoid using `TensorParallelMultiAdapterLinear` in any special case (phi3 and baichuan)

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py]: same as above, `TensorParallelMultiAdapterLinear` is avoided for special cases

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py]: unused and removed!

File: server/Makefile-lorax-punica
Comment (danieldk) [reply server/Makefile-lorax-punica:1]: Ah yeah, they have releases, so I kinda assumed that they upload wheels to PyPI.

Comment (drbh) [reply server/Makefile-lorax-punica:1]: yea unfortunately it looks like only the client is published to pypi

File: server/text_generation_server/models/globals.py
Comment (danieldk) [inline server/text_generation_server/models/globals.py:33]:
```suggestion
ADAPTER_TO_INDEX: Dict[str, int] = None
```

(Not sure if it's the correct type.)

File: server/text_generation_server/models/globals.py
Comment (danieldk) [inline server/text_generation_server/models/globals.py:36]:
```suggestion
def set_adapter_to_index(adapter_to_index: Dict[str, int]):
```

File: server/text_generation_server/models/model.py
Comment (danieldk) [inline server/text_generation_server/models/model.py:276]: It feels like this should be in `server/text_generation_server/adapters/lora.py` maybe? (along with the `shard_on_dim` helper)

File: server/text_generation_server/models/model.py
Comment (danieldk) [inline server/text_generation_server/models/model.py:296]: Unused?

File: server/text_generation_server/utils/merges/strategies.py
Comment (danieldk) [inline server/text_generation_server/utils/merges/strategies.py:1]: Probably worth adding a header here too, since it's mostly from https://github.com/predibase/lorax/blob/49bb52f357863974be100da19cc947f504c18d77/server/lorax_server/utils/merges/strategies.py#L43 ?

**Comment by danieldk** [comment]
Forgot to add: we probably want an integration test as well.

File: server/text_generation_server/layers/lora.py
Comment (drbh) [reply server/text_generation_server/layers/lora.py:29]: thanks! updated

File: server/text_generation_server/layers/lora.py
Comment (drbh) [reply server/text_generation_server/layers/lora.py:158]: updated

File: server/text_generation_server/models/model.py
Comment (drbh) [reply server/text_generation_server/models/model.py:296]: removed

File: server/text_generation_server/utils/merges/strategies.py
Comment (drbh) [reply server/text_generation_server/utils/merges/strategies.py:1]: great catch. updated

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py:90]: yea currently an exception is thrown if the model does not support adapters 
```python
raise ValueError("This model does not support adapter loading.")
```
and if the lora is not compatible with the model type we'll also throw
```python
raise ValueError(
    f"Adapter '{adapter_id}' is not compatible with model '{model_id}'. "
    f"Architectures differ: {model_config.architectures} != {expected_config.architectures}. "
    f"Use --model-id '{adapter_config.base_model_name_or_path}' instead."
)
```

File: server/text_generation_server/layers/lora.py
Comment (drbh) [reply server/text_generation_server/layers/lora.py]: ah yes we should throw if either `lora_a_ptr` OR `lora_b_ptr` is None. I've moved the check above the `if data.use_sgmv:` condition and throw is either are missing

File: server/text_generation_server/models/globals.py
Comment (drbh) [reply server/text_generation_server/models/globals.py]: thats correct and updated!

File: server/text_generation_server/models/globals.py
Comment (drbh) [reply server/text_generation_server/models/globals.py]: updated

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py]: updated so `TensorParallelMultiAdapterLinear` and `TensorParallelAdapterRowLinear` explicitly no-op unless the server was started with adapters

File: server/text_generation_server/layers/lora.py
Comment (drbh) [reply server/text_generation_server/layers/lora.py:210]: my understanding is that `start_idx` and `end_index` are used to indicate segments in a tensor that each need to have the adapters applied. For example 

in this case

```python
sizes = [
    head_size * config.num_attention_heads,
    head_size * config.num_key_value_heads,
    head_size * config.num_key_value_heads,
]
base_layer = TensorParallelColumnLinear.load_multi(
    config,
    prefixes=[f"{prefix}.q_proj", f"{prefix}.k_proj", f"{prefix}.v_proj"],
    dim=0,
    weights=weights,
    bias=bias,
)

TensorParallelMultiAdapterLinear.load(
    base_layer=base_layer,
    layer_id=layer_id,
    layer_names=["q_proj", "k_proj", "v_proj"],
    sizes=sizes,
    process_group=weights.process_group,
)
```

the `start_idx` and `end_idx` will be based on the `sizes` defined for the 3 tensors (num_attention_heads for q, and num_key_value_heads for k and v). The code above (in `lora.py`) will iterate over those three segments and apply the adapter in place in the `result`. In the `TensorParallelAdapterRowLinear` the case, the layers refer to segments like `gate_proj` and `up_proj`... not 100% sure if that's correct, but is my current understanding.

File: server/text_generation_server/layers/lora.py
Comment (drbh) [reply server/text_generation_server/layers/lora.py:270]: I believe the intention is to update the result in place using the indices. We may be able to pre slice and still update in place, however maybe we can revisit this improvement in a future PR? what do you think?

File: server/text_generation_server/models/model.py
Comment (drbh) [reply server/text_generation_server/models/model.py]: agreed, moved both!

File: server/text_generation_server/layers/lora.py
Comment (danieldk) [reply server/text_generation_server/layers/lora.py:63]: Thanks for the explanation!

File: server/text_generation_server/layers/lora.py
Comment (danieldk) [reply server/text_generation_server/layers/lora.py:270]: Yeah, sounds good.

**Review by danieldk — APPROVED** [review]
Thanks for all the changes! Looks ready to merge to me after the small nit that breaks CI is fixed.

**Comment by drbh** [comment]
@danieldk thanks for the review! I've fixed the nits and CI passes. Going to go ahead and merge based on your last approval

**Comment by mhou7712** [comment]
Hi @flozi00, is it possible to look at my issue https://github.com/huggingface/text-generation-inference/issues/2143 and let me know any suggestion?  Thanks.

**Pull request merged**
